### DISCUSSION TITLE
How to get huma.Context from context.Context

### DISCUSSION THREAD

**Discussion by davidolrik** [danielgtaylor/huma#410] [Q&A] [answered]
When I create a handler using huma.Register the handler signature is  `func(ctx context.Context, input *struct{}) (output *struct{}, error)`.

How do I get the the huma.Context from that `ctx`?

**Comment by x-user** [comment]
You can pass it via context from middleware, but it against huma design principle, so no builtin way to do this.

**Comment by danielgtaylor** [comment]
@davidolrik what are you trying to accomplish using the `huma.Context`? There may be a more idiomatic way to do it, but as stated above you can inject it via middleware or resolver if absolutely needed.

**Reply by davidolrik** [reply]
I wanted access to the underlying http request to get the client ip.

I did create a middleware that sets the `huma.Context` in the context but it seems like a hack, so I ended up just getting the client ip and adding it to the context. (This also seems like a hack as it is done for every request, and not just the one that needs it)

**Reply by x-user** [reply]
I wanted to say that you can use resolver to make it request specific, but there no way to get IP from `huma.Context`. And even if we add this functionality it will not work in fiber (https://github.com/danielgtaylor/huma/issues/345).

**Reply by davidolrik** [reply]
I ended up creating a very slim middleware, that I selectively apply to the endpoints that need it as part of the endpoint registration.

**Reply by joa23** [reply]
@davidolrik - how do you get access to the http.request from huma.Context? The underlaying e.g. chiContext is not exported so I struggle to find a way. Thanks for any pointers.

**Reply by davidolrik** [reply, marked as answer]
> @davidolrik - how do you get access to the http.request from huma.Context? 

What I did, was creating a small middleware that added the client ip from request to the context like this:

```go
type CtxClientIP struct{}

func HumaMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		r = r.WithContext(context.WithValue(r.Context(), CtxClientIP{}, r.RemoteAddr))
		next.ServeHTTP(w, r)
	})
}
```

You can then access the client ip in your request handlers like so:

```go
ip := ctx.Value(CtxClientIP{}).(string)
```

If you need access to the whole request, then just put `r` in to the context
